[WTF] Optimize tryGetUtf8ForRange by introducing SIMD find8NonASCII

Constellation opened the pull request:
#### 1b93c3178bdb088707ca358225f7dcf0f58ea5f4
<pre>
[WTF] Optimize tryGetUtf8ForRange by introducing SIMD find8NonASCII
<a href="https://bugs.webkit.org/show_bug.cgi?id=243925">https://bugs.webkit.org/show_bug.cgi?id=243925</a>

Reviewed by Darin Adler.

This patch introduces find8NonASCII and find16NonASCII. Later, both can be used in isAllASCII implementation.
We use it for tryGetUtf8ForRange improvement by avoiding allocating temporary Vector. We also extract it as
a template so that we can avoid allocating temporary CString in TextEncoder#encode.
The microbenchmark showed TextEncoder#encode improvement by 15%.

* Source/WTF/wtf/text/StringCommon.cpp:
(WTF::find8NonASCIIAlignedImpl):
(WTF::find16NonASCIIAlignedImpl):
* Source/WTF/wtf/text/StringCommon.h:
(WTF::find8NonASCII):
(WTF::find16NonASCII):
* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::StringImpl::utf8ForCharacters):
(WTF::StringImpl::tryGetUtf8ForRange const):
* Source/WTF/wtf/text/StringImpl.h:
(WTF::StringImpl::tryGetUtf8ForRange const):
* Source/WebCore/dom/TextEncoder.cpp:
(WebCore::TextEncoder::encode const):

Canonical link: <a href="https://commits.webkit.org/253410@main">https://commits.webkit.org/253410@main</a>
</pre>

darinadler commented on Source/WTF/wtf/text/StringImpl.h:
```diff
@@ -1354,6 +1358,62 @@ inline Ref<StringImpl> StringImpl::createByReplacingInCharacters(const UChar* ch
     return newImpl;
 }
 
+template<typename Func>
```
Is there a reason we use `func` instead of `function`?

Constellation replied:
Changed.

darinadler reviewed (approved):
I see the microbenchmark improvement. Did you measure any effect on larger benchmarks?

Constellation committed:
[WTF] Optimize tryGetUtf8ForRange by introducing SIMD find8NonASCII
https://bugs.webkit.org/show_bug.cgi?id=243925

Reviewed by Darin Adler.

This patch introduces find8NonASCII and find16NonASCII. Later, both can be used in isAllASCII implementation.
We use it for tryGetUtf8ForRange improvement by avoiding allocating temporary Vector. We also extract it as
a template so that we can avoid allocating temporary CString in TextEncoder#encode.
The microbenchmark showed TextEncoder#encode improvement by 15%.

* Source/WTF/wtf/text/StringCommon.cpp:
(WTF::find8NonASCIIAlignedImpl):
(WTF::find16NonASCIIAlignedImpl):
* Source/WTF/wtf/text/StringCommon.h:
(WTF::find8NonASCII):
(WTF::find16NonASCII):
* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::StringImpl::utf8ForCharacters):
(WTF::StringImpl::tryGetUtf8ForRange const):
* Source/WTF/wtf/text/StringImpl.h:
(WTF::StringImpl::tryGetUtf8ForRange const):
* Source/WebCore/dom/TextEncoder.cpp:
(WebCore::TextEncoder::encode const):

Canonical link: https://commits.webkit.org/253410@main
